File: NewsApp/app/src/main/java/com/example/shakil/newsapp/Adapter/ListNewsAdapter.java
package com.example.shakil.newsapp.Adapter;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.shakil.newsapp.Common.ISO8601Parse;
import com.example.shakil.newsapp.DetailsArticle;
import com.example.shakil.newsapp.Interface.ItemClickListener;
import com.example.shakil.newsapp.ListNews;
import com.example.shakil.newsapp.Model.Article;
import com.example.shakil.newsapp.R;
import com.github.curioustechizen.ago.RelativeTimeTextView;
import com.squareup.picasso.Picasso;

import java.text.ParseException;
import java.util.Date;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;

public class ListNewsAdapter extends RecyclerView.Adapter<ListNewsAdapter.ListNewsViewHolder> {

    private Context context;
    private List<Article> articleList;

    public ListNewsAdapter(Context context, List<Article> articleList) {
        this.context = context;
        this.articleList = articleList;
    }

    @NonNull
    @Override
    public ListNewsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(context).inflate(R.layout.news_layout, parent, false);
        return new ListNewsViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ListNewsViewHolder holder, int position) {
        Picasso.get().load(articleList.get(position).getUrlToImage()).into(holder.article_image);

        if (articleList.get(position).getTitle().length() > 65) {
            holder.article_title.setText(articleList.get(position).getTitle().substring(0, 65) + "...");
        }
        else {
            holder.article_title.setText(articleList.get(position).getTitle());
        }

        if (articleList.get(position).getPublishedAt() != null){
            Date date = null;
            try {
                date = ISO8601Parse.parse(articleList.get(position).getPublishedAt());
            } catch (ParseException e) {
                e.printStackTrace();
            }
            holder.article_time.setReferenceTime(date.getTime());
        }

        holder.setItemClickListener(new ItemClickListener() {
            @Override
            public void onClick(View view, int position) {
                Intent intent = new Intent(context, DetailsArticle.class);
                intent.putExtra("webURL", articleList.get(position).getUrl());
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                context.startActivity(intent);
            }
        });
    }

    @Override
    public int getItemCount() {
        return articleList.size();
    }

    public class ListNewsViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        TextView article_title;
        RelativeTimeTextView article_time;
        CircleImageView article_image;

        ItemClickListener itemClickListener;

        public void setItemClickListener(ItemClickListener itemClickListener) {
            this.itemClickListener = itemClickListener;
        }

        public ListNewsViewHolder(@NonNull View itemView) {
            super(itemView);

            article_title = itemView.findViewById(R.id.article_title);
            article_time = itemView.findViewById(R.id.article_time);
            article_image = itemView.findViewById(R.id.article_image);

            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {
            itemClickListener.onClick(v, getAdapterPosition());
        }
    }
}

File: KotlinNewsApp/app/src/main/java/com/example/shakil/kotlinnewsapp/Adapter/ListNewsAdapter.kt
package com.example.shakil.kotlinnewsapp.Adapter

import android.content.Context
import android.content.Intent
import android.hardware.Camera
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.shakil.kotlinnewsapp.Common.ISO8601DateParser
import com.example.shakil.kotlinnewsapp.Interface.ItemClickListener
import com.example.shakil.kotlinnewsapp.ListNewsActivity
import com.example.shakil.kotlinnewsapp.Model.Article
import com.example.shakil.kotlinnewsapp.Model.WebSite
import com.example.shakil.kotlinnewsapp.NewsDetailsActivity
import com.example.shakil.kotlinnewsapp.R
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.news_layout.view.*
import kotlinx.android.synthetic.main.source_news_layout.view.*
import java.text.ParseException
import java.util.*

class ListNewsAdapter(val context: Context, val articleList: MutableList<Article>):
    RecyclerView.Adapter<ListNewsAdapter.ListNewsViewHolder>() {

    class ListNewsViewHolder(itemView: View): RecyclerView.ViewHolder(itemView), View.OnClickListener {
        private lateinit var itemClickListener: ItemClickListener

        var article_title = itemView.article_title
        var article_time = itemView.article_time
        var article_image = itemView.article_image
        init {
            itemView.setOnClickListener(this)
        }

        fun setItemClickListener(itemClickListener: ItemClickListener) {
            this.itemClickListener = itemClickListener
        }

        override fun onClick(v: View?) {
            itemClickListener.onClick(v!!, adapterPosition)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListNewsViewHolder {
        return ListNewsViewHolder(LayoutInflater.from(context).inflate(R.layout.news_layout, parent, false))
    }

    override fun getItemCount(): Int {
        return articleList.size
    }

    override fun onBindViewHolder(holder: ListNewsViewHolder, position: Int) {
        /*Image Load*/
        Picasso.get().load(articleList[position].urlToImage).into(holder.article_image)

        if (articleList[position].title!!.length > 65) {
            holder.article_title.text = articleList[position].title!!.substring(0, 65) + "..."
        }
        else {
            holder.article_title.text = articleList[position].title!!
        }

        if (articleList[position].publishedAt != null) {
            var date: Date? = null
            try {
                date = ISO8601DateParser.parse(articleList[position].publishedAt!!)
            } catch (ex: ParseException) {
                ex.printStackTrace()
            }
            holder.article_time.setReferenceTime(date!!.time)
        }

        holder.setItemClickListener(object : ItemClickListener {
            override fun onClick(view: View, position: Int) {
                val details = Intent(context, NewsDetailsActivity::class.java)
                details.putExtra("webURL", articleList[position].url)
                details.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                context.startActivity(details)
            }
        })
    }
}